perf(core): avoid pulling in jit-specific code in aot bundles

crisbeto opened the pull request:
In #29083 a call to `getCompilerFacade` was added to `ApplicationRef` which pulls in a bit of JIT-specific code. Since the code path that calls the function can't be hit for an AOT-compiled app, these changes add an `ngJitMode` guard which will allow for dead code elimination to drop it completely. Testing it out against a new CLI project showed a difference of ~1.2kb.

devversion commented on packages/core/src/application_ref.ts:
```diff
@@ -86,7 +86,7 @@ export function compileNgModuleFactory__POST_R3__<M>(
   // modules can be still passed through "bootstrapModule". In that case we shouldn't
   // unnecessarily require the JIT compiler.
-  if (compilerProviders.length === 0) {
+  if (compilerProviders.length === 0 || (typeof ngJitMode !== 'undefined' && !ngJitMode)) {
```
Out of curiosity: Did you consider doing this before `isComponentResourceResolutionQueueEmpty`? That might avoid bringing in the `isComponentResourceResolutionQueueEmpty` function?

JoostK replied:
Yep, I think we can do better than this. The `compilerOptions`/`compilerProviders` is not used in AOT mode, so the JIT guard could span more code.

crisbeto replied:
Nice 👍 . Changed.

JoostK reviewed (approved):
💯

IgorMinar commented on goldens/size-tracking/integration-payloads.json:
```diff
@@ -61,8 +61,8 @@
-        "bundle": "TODO(i): after removal of tsickle from ngc-wrapped / ng_package, we had to switch to SIMPLE optimizations which increased the size from 252338 to 1197869, see PR#37221 and PR#37317 for more info",
-        "bundle": 1197869
+        "bundle": "TODO(i): after removal of tsickle from ngc-wrapped / ng_package, we had to switch to SIMPLE optimizations which increased the size from 252338 to 1197896, see PR#37221 and PR#37317 for more info",
+        "bundle": 1197896
```
isn't it odd that this one went up? not a blocker, just weird.

crisbeto replied:
This one was also smaller the first time around, but I had to resolve a merge conflict.

IgorMinar reviewed (approved):
LGTM! Thank you!

Reviewed-for: global-approvers

crisbeto committed:
perf(core): avoid pulling in jit-specific code in aot bundles

In #29083 a call to `getCompilerFacade` was added to `ApplicationRef` which pulls in a bit of JIT-specific code. Since the code path that calls the function can't be hit for an AOT-compiled app, these changes add an `ngJitMode` guard which will allow for dead code elimination to drop it completely. Testing it out against a new CLI project showed a difference of ~1.2kb.

atscott commented:
@crisbeto As discussed, switching this to master only as there are conflicts with the patch branch. Please create a follow-up for patch only
